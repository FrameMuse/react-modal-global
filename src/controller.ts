/*

MIT License

Copyright (c) 2023 Valery Zinchenko

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

*/

import EventEmitter from "eventemitter3"

import { ModalComponent, ModalParams, ModalWindow, ModalWindowParams } from "./types"
import { serialize } from "./utils"

const DEFAULT_PARAMS: ModalParams = {
  id: 0,
  closable: true,
  weak: false,
  fork: false
}

export interface ModalState {
  isOpen: boolean
  windows: ModalWindow[]
}

interface Events {
  add: [ModalWindow]
  remove: [ModalWindow]
  update: []
}

/**
 * Controller for opening and closing modal windows.
 *
 * Can be used with `ModalContainer` or with custom implementation.
 */
export class ModalController {
  #isOpen = false
  protected set isOpen(value: boolean) {
    this.#isOpen = value
  }
  /**
   * Whether any modals are shown.
   */
  public get isOpen(): boolean {
    return this.#isOpen
  }

  protected windows: Set<ModalWindow> = new Set
  protected events: EventEmitter<Events> = new EventEmitter


  /**
   * Opens a modal window. Infers props from the component.
   *
   * - If the same modal window is already in the queue, it will be ignored.
   * - If the modal window is weak, it will be unmounted after closing.
   * - If the modal window is forked, it will be opened over all other modals.
   * - If the modal window is closable (be default `true`), it will be closed by clicking on the overlay.
   * - If the modal window is not closable, it will be closed only by calling internal `close` method.
   *
   * @param component Modal component.
   * @param params Modal params.
   * @returns Modal window and `PromiseLike`.
   *
   * @example
   * const modal = modalController.open(MyModal, { id: 1 })
   * modal.closable // `true` by default
   *
   * modal.then(() => console.log("Modal was closed"))
   * modal.close()
   */
  public open<P>(component: ModalComponent<P>, ...[modalParams]: ModalWindowParams<P>): ModalWindow<P> & PromiseLike<void> {
    let resolveFunction = () => { /* Noop */ }
    const promise = new Promise<void>(resolve => resolveFunction = resolve)
    const close = () => {
      this.remove(modal as ModalWindow<unknown>)
      resolveFunction()
    }

    const params: ModalParams & P = { ...DEFAULT_PARAMS, ...modalParams as P }
    const modal: ModalWindow<P> = { component, params, close, closed: false, focused: true }

    this.add(modal as ModalWindow<unknown>)

    return {
      ...modal,
      then(onfulfilled?, onrejected?) {
        return promise.then(onfulfilled, onrejected)
      },
    }
  }
  /**
   * Replaces the last modal window in the queue with a new one.
   *
   * If the queue is empty, it will be added to the queue.
   */
  public replace<P>(component: ModalComponent<P>, ...[params]: ModalWindowParams<P>): ModalWindow<P> & PromiseLike<void> {
    const lastWindow = [...this.windows].at(-1)
    if (lastWindow != null) {
      this.windows.delete(lastWindow)
    }

    return this.open(component, params as never)
  }

  /**
   * Adds a modal window to the queue.
   *
   * - Controls whether the modal is open.
   * - Controls whether the order of windows.
   */
  private add(modalWindow: ModalWindow) {
    // If there are temporary modal windows, clear.
    if (this.isOpen === false && this.windows.size > 0) {
      this.windows.clear()
    }

    this.isOpen = true
    this.windows.add(modalWindow)
    this.events.emit("add", modalWindow)
  }
  /**
   * Removes a modal window from the queue.
   */
  private remove(modalWindow: ModalWindow) {
    if (!this.windows.has(modalWindow)) return

    const isLastWindow = this.windows.size === 1
    if (isLastWindow) {
      this.isOpen = false

      // If the modal window is weak, it will be unmounted after closing.
      // Otherwise, skip removing to let last modal window to be in the queue.
      if (!modalWindow.params.weak) {
        this.events.emit("update")
        return
      }
    }

    this.windows.delete(modalWindow)
    this.events.emit("remove", modalWindow)
  }

  private findByComponent<Params>(component: ModalComponent<Params>, params?: ModalParams & Params): ModalWindow<Params>[] {
    const foundWindows = [...this.windows].filter(modal => {
      if (modal.component !== component) {
        return false
      }

      if (params != null) {
        return serialize(params) === serialize(modal.params)
      }

      return true
    })
    return foundWindows as ModalWindow<Params>[] // Assume that found windows follow the params type (`Params`).
  }

  private findById(id: ModalParams["id"]): ModalWindow[] {
    const foundWindows = [...this.windows].filter(modal => modal.params.id === id)
    return foundWindows
  }

  /**
   * Closes all modals by its component (including forked) starting from the last one.
   */
  public closeByComponent<Params>(component: ModalComponent<Params>, params?: ModalParams & Params) {
    this.findByComponent(component, params).forEach(modal => modal.close())
  }
  /**
   * Closes all modals by its id (including forked) starting from the last one.
   */
  public closeById(id: ModalParams["id"]) {
    this.findById(id).forEach(modal => modal.close())
  }
  /**
   * Closes all modals (including forked).
   */
  public closeAll() {
    this.isOpen = false
    this.events.emit("update")
    // All windows are now treated as temporary and they will be removed on the next `add`.
  }

  /**
   * Subscribes on event.
   *
   * @returns `unsubscribe` method
   */
  public observe(callback: (state: ModalState) => void) {
    const listener = () => {
      const state: ModalState = {
        isOpen: this.isOpen,
        windows: [...this.windows]
      }
      callback(state)
    }

    this.events.on("add", listener)
    this.events.on("remove", listener)
    this.events.on("update", listener)

    this.events.emit("update") // Emit initial state

    return () => {
      this.events.off("add", listener)
      this.events.off("remove", listener)
      this.events.off("update", listener)
    }
  }
}


export const Modal = new ModalController
