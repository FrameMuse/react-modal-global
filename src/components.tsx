import _ from "lodash"
import { ReactNode, useEffect, useId } from "react"
import { useUnmount } from "react-use"

import { Modal, ModalController } from "./controller"
import { ModalComponent, ModalParams } from "./types"
import { serialize } from "./utils"


interface ModalViewProps {
  controller?: ModalController

  children: ReactNode
}

export function ModalView(props: ModalViewProps) {
  const id = useId()

  useEffect(() => {
    function ModalViewComponent() {
      return <>{props.children}</>
    }

    const controller = props.controller || Modal
    controller.open(ModalViewComponent, { id })
  }, [props.children, props.controller])

  useUnmount(() => {
    const controller = props.controller || Modal
    controller.closeById(id)
  })

  return null
}





interface ModalOpenProps<ComponentProps> extends Partial<ModalParams> {
  component: ModalComponent<ComponentProps>
  componentProps: ComponentProps

  controller?: ModalController
}

export function ModalOpen<ComponentProps>(props: ModalOpenProps<ComponentProps>) {
  useEffect(() => {
    const controller = props.controller || Modal
    const params = { ...props.componentProps, ..._.omit(props, "component", "componentProps", "controller") }
    const modal = controller.open(props.component, params)


    return () => {
      modal.close()
    }
  }, [props.component, serialize(props.componentProps), props.controller])

  return null
}
