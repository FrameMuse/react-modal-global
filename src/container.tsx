import React from "react"
import { classWithModifiers } from "./utils"
import { PopupContext } from "./context"
import { PopupWindow } from "./interfaces"
import { PopupPrivate } from "./controller"

export interface PopupContainerProps {
  className?: string
}
export interface PopupContainerState {
  isActive: boolean
  queue: PopupWindow[]
}

export class PopupContainer extends React.Component<PopupContainerProps, PopupContainerState> {
  state: PopupContainerState = {
    isActive: false,
    queue: []
  }

  constructor(props: any) {
    super(props)
    // Set Popup dispatcher
    PopupPrivate.dispatch = this.setState.bind(this)
  }

  render() {
    const { isActive, queue } = this.state
    const lastPopup = queue[queue.length - 1]
    const { component: PopupWindowComponent, params = {}, close } = lastPopup || {}

    const className = this.props.className || "popup"
    return (
      <div className={classWithModifiers(className, isActive && "active")}>
        <div className={className + "__container"} onClick={close}>
          <div className={className + "__inner"} onClick={event => event.stopPropagation()}>
            <PopupContext.Provider value={lastPopup}>
              {PopupWindowComponent && <PopupWindowComponent {...params} />}
            </PopupContext.Provider>
          </div>
        </div>
      </div>
    )
  }
}
