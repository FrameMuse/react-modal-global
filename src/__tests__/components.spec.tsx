import "@testing-library/jest-dom"

import { render } from "@testing-library/react"

import { ModalOpen,ModalView } from "../components"
import { ModalContainer } from "../container.new"
import { ModalController } from "../controller"

let Modal: ModalController = new ModalController()

beforeEach(() => {
  Modal = new ModalController()
})

describe("Components", () => {
  test("ModalView", () => {
    const { container } = render(
      <>
        <ModalContainer controller={Modal} />
        <ModalView controller={Modal}>
          <div id="test1">123</div>
          <div id="test2">675</div>
        </ModalView>
      </>
    )

    expect(container).toMatchSnapshot()

    expect(container.querySelector("#test1")).toHaveTextContent("123")
    expect(container.querySelector("#test2")).toHaveTextContent("675")
    expect(container).toHaveTextContent("123675")
  })

  test("ModalOpen", () => {
    function Test(props: { a: 1 }) { return <div id="a">a{props.a}</div> }

    const { container } = render(
      <>
        <ModalContainer controller={Modal} />

        <ModalOpen
          component={Test}
          componentProps={{ a: 1 }}

          controller={Modal}
        />
      </>
    )

    expect(container).toMatchSnapshot()
    expect(container.querySelector("#a")).toHaveTextContent("a1")
  })
})
