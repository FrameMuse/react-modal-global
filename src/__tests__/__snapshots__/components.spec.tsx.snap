// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components ModalOpen 1`] = `
<div>
  <div
    aria-hidden="false"
    aria-modal="true"
    class="modal modal--active"
  >
    <div
      class="modal__container"
    >
      <div
        id="a"
      >
        a
        1
      </div>
    </div>
  </div>
</div>
`;

exports[`Components ModalView 1`] = `
<div>
  <div
    aria-hidden="false"
    aria-modal="true"
    class="modal modal--active"
  >
    <div
      class="modal__container"
    >
      <div
        id="test1"
      >
        123
      </div>
      <div
        id="test2"
      >
        675
      </div>
    </div>
  </div>
</div>
`;
